### PR TITLE
Linux: determine ProcessName using /proc cmdline to avoid truncated names

### PR DESCRIPTION
**Pull request by tmds**
Fixes https://github.com/dotnet/corefx/issues/34437

@stephentoub @danmosemsft @krwq @wtgodbe ptal

### CODE REVIEW

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs [hunk 1]
```diff
@@ -256,6 +255,73 @@ internal static string GetExePath(int processId = -1)
...
+                        {
+                            uint newLength = (uint)buffer.Length * 2;
+
+                            byte[] tmp = ArrayPool<byte>.Shared.Rent((int)newLength);
+                            buffer.CopyTo(tmp);
+                            if (rentedArray != null)
+                            {
+                                ArrayPool<byte>.Shared.Return(rentedArray);
+                            }
+                            buffer = rentedArray = tmp;
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.Linux.cs]: Nit: I'd prefer it if the code were instead something like:
```C#
byte[] toReturn = rentedArray;
buffer = rentedArray = tmp;
if (toReturn != null)
{
    ArrayPool<byte>.Shared.Return(toReturn);
}
```
The concern is if something were to go wrong (e.g. an OOM) resulting in an exception after the rented array was stored into the pool but before rentedArray was updated to point to a different buffer, then that exception would traverse the finally block, which would return the same buffer to the pool again, resulting in the single buffer being in the pool twice.

File: src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs [hunk 2]
```diff
@@ -117,21 +117,21 @@ internal static ProcessInfo CreateProcessInfo(int pid, ReusableTextReader reusab
...
-        internal static ProcessInfo CreateProcessInfo(Interop.procfs.ParsedStat procFsStat, ReusableTextReader reusableReader)
+        internal static ProcessInfo CreateProcessInfo(Interop.procfs.ParsedStat procFsStat, ReusableTextReader reusableReader, string processName)
         {
             int pid = procFsStat.pid;
 
             var pi = new ProcessInfo()
             {
                 ProcessId = pid,
-                ProcessName = procFsStat.comm,
+                ProcessName = processName ?? Process.GetProcessName(pid) ?? string.Empty,
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Linux.cs]: What kind of throughput impact does this have on Process.GetProcesses()?  Anything measurable?

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 3]
```diff
@@ -1875,6 +1875,37 @@ public void TestLongProcessIsWorking()
             Assert.True(p.HasExited);
         }
 
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        [ActiveIssue(37054, TestPlatforms.OSX)]
+        [Fact]
+        public void LongProcessNamesAreSupported()
+        {
+            string commandName = "sleep";
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: Nit: const string CommandName

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 4]
```diff
@@ -1875,6 +1875,37 @@ public void TestLongProcessIsWorking()
...
+
+            if (!IsProgramInstalled(commandName))
+            {
+                return;
+            }
+
+            string longProcessName = "123456789012345678901234567890";
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: Nit: const string LongProcessName

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 5]
```diff
@@ -1875,6 +1875,37 @@ public void TestLongProcessIsWorking()
...
         }
 
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
+        [ActiveIssue(37054, TestPlatforms.OSX)]
+        [Fact]
+        public void LongProcessNamesAreSupported()
+        {
+            string commandName = "sleep";
+
+            if (!IsProgramInstalled(commandName))
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: Nit:
```C#
string programPath = GetProgramPath(commandName);
if (programPath == null)
{
    return;
}

... // then use programPath instead of GetProgramPath below
```
?

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 6]
```diff
@@ -1875,6 +1875,37 @@ public void TestLongProcessIsWorking()
...
+                return;
+            }
+
+            string longProcessName = "123456789012345678901234567890";
+            string sleepCommandPathFileName = Path.Combine(TestDirectory, longProcessName);
+            File.Copy(GetProgramPath(commandName), sleepCommandPathFileName);
+
+            using (Process px = Process.Start(sleepCommandPathFileName, "30"))
+            {
+                var runningProcesses = Process.GetProcesses();
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: Nit: var => Process[]

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 7]
```diff
@@ -1875,6 +1875,37 @@ public void TestLongProcessIsWorking()
...
+            if (!IsProgramInstalled(commandName))
+            {
+                return;
+            }
+
+            string longProcessName = "123456789012345678901234567890";
+            string sleepCommandPathFileName = Path.Combine(TestDirectory, longProcessName);
+            File.Copy(GetProgramPath(commandName), sleepCommandPathFileName);
+
+            using (Process px = Process.Start(sleepCommandPathFileName, "30"))
```
Comment (danmosemsft) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: Since we expect to kill this process immediatley we are done with it, it could wait essentially forever instead of 30 seconds. That would avoid the case where the machines are so overloaded or unlucky that we don't read all the processes before the sleep expires. Put another way, this 30 seconds does not affect the length of the test.

**Comment by tmds** [comment]
@stephentoub @danmosemsft thank you for reviewing, I've addressed the feedback.

File: src/System.Diagnostics.Process/tests/ProcessTests.cs [hunk 8]
```diff
@@ -1875,6 +1875,32 @@ public void TestLongProcessIsWorking()
...
+            string sleepCommandPathFileName = Path.Combine(TestDirectory, LongProcessName);
+            File.Copy(programPath, sleepCommandPathFileName);
+
+            using (Process px = Process.Start(sleepCommandPathFileName, "600"))
+            {
+                Process[] runningProcesses = Process.GetProcesses();
+                Assert.Contains(runningProcesses, p => p.ProcessName == LongProcessName);
+
+                px.Kill();
+                px.WaitForExit();
```
Comment (stephentoub) [inline src/System.Diagnostics.Process/tests/ProcessTests.cs]: Why did you remove the try/finally you originally had here?

Comment (tmds) [reply src/System.Diagnostics.Process/tests/ProcessTests.cs]: I had copied it from the test that was in the other PR. I removed it because it should never throw.

Comment (stephentoub) [reply src/System.Diagnostics.Process/tests/ProcessTests.cs]: It will if the assert fails.

Comment (tmds) [reply src/System.Diagnostics.Process/tests/ProcessTests.cs]: Ah, yes. Let me add it back.

**Comment by stephentoub** [comment]
Thanks.
